Add a Factory for building blocks and Vectors

ChrisHegarty opened the pull request:
This commit adds a BlockFactory - an extra level of indirection when building blocks. The factory couples circuit breaking when building, allowing for incrementing the breaker as blocks and Vectors are built.

This PR adds the infrastructure to allow us to move the operators and implementations over to the factory, rather than actually moving all there at once.

Note: we still need to fix up many tests and operators, but much of this can come later.

relates #99173

ChrisHegarty committed:
Add an initial Block Factory

elasticsearchmachine commented:
Pinging @elastic/es-ql (Team:QL)

elasticsearchmachine commented:
Pinging @elastic/elasticsearch-esql (:Query Languages/ES|QL)

ChrisHegarty committed:
fix benchmark builds

ChrisHegarty committed:
Merge branch 'main' into blockFactory

ChrisHegarty commented:
Actually, maybe we make a BlockFactory with a noop breaker the default (rather than one with a real breaker), this would allow to migrate gradually to the real breaking factory.

ChrisHegarty commented:
Actually, the default for now can be a no-op breaking blockBuilder, which is effective opt-out, then allow use-sites opt-in. When all are opted-in, we can flip the default from non-breaking to breaking.

ChrisHegarty committed:
use a non-breaking block factory by default

ChrisHegarty committed:
revert changes in block hash, for now

ChrisHegarty committed:
add tests and fixes

ChrisHegarty committed:
Merge branch 'main' into blockFactory

ChrisHegarty committed:
update block factory tests

ChrisHegarty committed:
tests

nik9000 commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/BlockFactory.java:
```diff
@@ -0,0 +1,265 @@
+    );
+
+    private static final BlockFactory GLOBAL = loadGlobalFactory();
+    // new BlockFactory(new NoopCircuitBreaker("esql_noop_breaker"), BigArrays.NON_RECYCLING_INSTANCE);
```
I think I'd prefer this not be static and live on `DriverContext`. We're already passing that in most places, right? In that case I don't think we need the `ServiceLoader` or anything like that - just pass the breaker in. Right?

ChrisHegarty replied:
Right.

nik9000 commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/ConstantNullBlock.java:
```diff
@@ -98,7 +100,7 @@ public Block expand() {
     @Override
     public long ramBytesUsed() {
-        return BASE_RAM_BYTES_USED;
+        return 0L;
```
Doesn't it still use the memory?

ChrisHegarty replied:
D'oh! yeah, it's marginal. I restored it.

nik9000 commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/X-BlockBuilder.java.st:
```diff
@@ -57,6 +62,11 @@ $endif$
 
+    @Override
+    protected int elementSize() {
+        return $if(BytesRef)$-1$else$$if(boolean)$Byte$elseif(int)$Integer$elseif(long)$Long$else$Double$endif$.BYTES$endif$;
```
I think we might already have a constant for this in the build.gradle file. Or we should add one.

ChrisHegarty replied:
Ah ha! yes. Fixed, see [00e708c](https://github.com/elastic/elasticsearch/pull/99657/commits/00e708cdc1b4a51d139c5e7cfdf2caf7018dc050)

nik9000 reviewed (commented):
I'd really prefer the factory be non-static. i think that'd make the flow around its dependencies a lot more obvious. I understand it requires a fair bit of plumbing though.

nik9000 commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/BlockFactory.java:
```diff
@@ -0,0 +1,265 @@
+
+    private final BigArrays bigArrays;
+
+    private final boolean checkBreaker = true;
```
Does this do anything actually? I think it'll always be true, right? Maybe we just remove it?

ChrisHegarty replied:
Removed. we can re-add it at some point if we decide that it could be useful. see [f766b1b](https://github.com/elastic/elasticsearch/pull/99657/commits/f766b1b26c0604887f7e02be98ff234a7d44319c)

nik9000 commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/BlockFactory.java:
```diff
@@ -0,0 +1,265 @@
+            // its' totals, so add without breaking
+            breaker.addWithoutBreaking(delta);
+        }
+    }
```
This is pretty reasonable. I kind of wish we didn't need the `isAlreadyCreated` bit, but we're sort of stuck with it. There are far worse things though.

ChrisHegarty replied:
👍 Yeah, we kinda need the flexibility to be able to create blocks in different ways.

nik9000 commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/BlockFactory.java:
```diff
@@ -0,0 +1,265 @@
+        var b = new LongArrayVector(values, positionCount, this);
+        adjustBreaker(b.ramBytesUsed(), true);
+        return b;
+    }
```
I imagine folks using this like:
```
blockFactory.adjustBreaker(positionCount * Long.BYTES, false);
long[] values = new long[positionCount];
for (int i = 0; i < positionCount; i++) {
  values[i] = doWhateverStuff
}
blockFactory.newLongArrayVectory(values, positionCount);
```

In a case like that we'd double count. Maybe in that case we'd want to call
```
blockFactory.newLongArrayVectory(values, positionCount, positionCount * Long.BYTES);
```
and let the builder take care of truing up the count with the reported `ramBytesUsed`.

This is probably fine to wait for a followup.

ChrisHegarty replied:
yeah, I think that creating Vectors (not blocks) will be a particularly popular. The vector builders should be good enough and efficient for this kinda usage, but I do think that this usage is important to support - pre-adjusting the breaker, similar to what you outlined. 

Maybe we can make this a little more friendly. Here is an API sketch.  

```
long preAdjustedBytes = blockFactory.preAdjustBreakerForLong(positionCount);
long[] values = new long[positionCount];
for (int i = 0; i < positionCount; i++) {
  values[i] = doWhateverStuff
}
blockFactory.newLongArrayVector(values, positionCount, preAdjustedBytes);
```

nik9000 replied:
Something like that! I wonder if maybe we should wrap this kind of thing in a `LongVector.FixedLengthBuilder`. Basically the same as a `LongVector.Builder` but without the expanding. It's cook down to this but more meaningful names. Either way, that's a problem for another PR.

nik9000 reviewed (commented):
I think this goes a long way in the right direction. I'd love to get it in and iterate. Though maybe without the static access and service loader. OTOH, even that can change after. And it'd be fun to try and integrate this into the evaluator generator. And, until I try to iterate on it myself, I don't really understand the fun around making sure we free everything.

I suppose the important thing is that we can migrate to using the factory from where we are piece by piece.

ChrisHegarty committed:
restore constant null block

ChrisHegarty committed:
simplify element size in template

ChrisHegarty committed:
remove unused checkBreaker

ChrisHegarty committed:
add pre-adjusted vector construction

ChrisHegarty commented:
> I think this goes a long way in the right direction. I'd love to get it in and iterate. Though maybe without the static access and service loader. OTOH, even that can change after. And it'd be fun to try and integrate this into the evaluator generator. And, until I try to iterate on it myself, I don't really understand the fun around making sure we free everything.
> 
> I suppose the important thing is that we can migrate to using the factory from where we are piece by piece.

I've addressed all the review comments so far, except this one -  the big one!! ;-)  First, I completely agree that we want to move to a place where the static's are no longer required - all blocks and vectors should be created through the BlockFactory of the driver, or in the case of exchanges an explicit BlockFactory that is given to the exchange service.

This PR allows both the work above and the updating of evaluators and operators to proceed separately, with some obvious overlap. How about we merge this PR as is, and then immediately followup, as above, so that the work can be somewhat parallelized.

nik9000 commented:
If you have plan for moving us away from the static access then I'm happy. Sometimes going from where you are to where you want to be is hard. That's ok. I'll have another look and hopefully approve and then we can talk later about your plan.

ChrisHegarty committed:
Merge branch 'main' into blockFactory

nik9000 commented on x-pack/plugin/esql/compute/src/main/generated-src/org/elasticsearch/compute/data/BooleanArrayVector.java:
```diff
@@ -80,6 +84,6 @@ public String toString() {
     @Override
     public void close() {
-        // no-op
+        super.close();
```
These `void foo() { super.foo(); }` calls always confuse me. Maybe just remove it?

ChrisHegarty replied:
I'll take a look at this.

ChrisHegarty replied:
Removed superfluous super calls

nik9000 commented on x-pack/plugin/esql/compute/src/main/generated-src/org/elasticsearch/compute/data/BooleanBlock.java:
```diff
@@ -157,12 +157,24 @@ static int hash(BooleanBlock block) {
     }
 
+    /** Returns a builder using the {@link BlockFactory#getNonBreakingInstance block factory}. */
+    // Eventually, this should use the GLOBAL breaking instance
```
I guess we'd remove this entirely when we remove the static access, right?

ChrisHegarty replied:
ah yes, I'll update the comment

nik9000 replied:
:+1:

ChrisHegarty replied:
Done

nik9000 commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/AbstractVectorBuilder.java:
```diff
@@ -10,22 +10,41 @@
             return;
         }
         int newSize = calculateNewArraySize(valuesLength);
+        // TODO: bytes ref
```
Is this TODO still live?

ChrisHegarty replied:
this todo should be removed.

ChrisHegarty replied:
Removed

nik9000 commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/BlockFactoryParameters.java:
```diff
@@ -0,0 +1,18 @@
+import org.elasticsearch.common.breaker.CircuitBreaker;
+import org.elasticsearch.common.util.BigArrays;
+
+public interface BlockFactoryParameters {
```
Maybe a note that we expect this to be somewhat temporary while we migrate away from static access.

nik9000 replied:
And probably a little bit about what it does in the mean time.

ChrisHegarty replied:
will do.

nik9000 commented on x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/CountingCircuitBreaker.java:
```diff
@@ -0,0 +1,69 @@
+
+import org.elasticsearch.common.breaker.CircuitBreaker;
+
+public class CountingCircuitBreaker implements CircuitBreaker {
```
I'd use `MockBigArrays.LimitedBreaker` with a limit of `1GB` or something. At least, that's what I've been doing and it seemed ok. Maybe it's overkill. But I needed it with my CONCAT PR.

ChrisHegarty replied:
That sounds good. I'll take a look

ChrisHegarty replied:
I used LimitBreaker of 1Gb, and removed CountCircuitBreaker.

nik9000 commented on x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/data/BasicBlockTests.java:
```diff
@@ -26,51 +36,57 @@
+    BigArrays bigArrays() {
+        var breakerService = mock(CircuitBreakerService.class);
+        when(breakerService.getBreaker(CircuitBreaker.REQUEST)).thenReturn(breaker);
+        return new MockBigArrays(PageCacheRecycler.NON_RECYCLING_INSTANCE, breakerService);
```
The `MockBigArrays` ctor that takes a `ByteSizeValue` gets you something pretty similar to this. There's a limit, but you can set it to 1GB and mostly ignore it.

ChrisHegarty replied:
ah, that might be more straightforward. I'll update.

ChrisHegarty replied:
I did not use MockBigArrays, as it creates a new instance of the limit breaker for each getBreaker on the breaker service call.  This seems like it could be a bug in MockBigArrays?  Well... it depends on the expected usage.

nik9000 replied:
> This seems like it could be a bug in MockBigArrays?

I think that's a bug, yeah. But not one to fix now.

nik9000 commented on x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/data/BasicBlockTests.java:
```diff
@@ -141,15 +157,17 @@ static void assertSingleValueDenseBlock(Block initialBlock) {
                 block = blockBuilder.build();
             } else {
-                block = new IntArrayVector(IntStream.range(0, positionCount).toArray(), positionCount).asBlock();
+                new IntArrayVector(IntStream.range(0, positionCount).toArray(), positionCount).asBlock();
```
Leftover?

ChrisHegarty replied:
leftover, will remove.

ChrisHegarty replied:
Removed

nik9000 commented on x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/data/BasicBlockTests.java:
```diff
@@ -851,9 +895,33 @@ private static <B extends Block, BB extends Block.Builder, T> void assertNullVal
+
+    static <T extends Releasable & Accountable> void releaseAndAssertBreaker(T data, CircuitBreaker breaker) {
+        Releasables.closeExpectNoException(data);
+        assertThat(breaker.getUsed(), is(0L));
```
I don't think you are using `Accountable` on `data` right now. Had you intended to assert that the breaker is >= to the size?

ChrisHegarty replied:
Oh yeah, I was asserting that > 0 - something was allocated, so the breaker getUsed should be positive.  I then modified the test to use a no-op breaker, so getUsed is always 0.  I think I'll revert and not use the no-op breaker, since I think that the > 0 is more useful.

ChrisHegarty replied:
I restored the > 0 assertion

nik9000 reviewed (commented):
I left a few more questions. It's *probably* worth one more round on the questions just so I know what direction you are going. And so I can help after this get's in.

nik9000 commented:
Once you are ready please merge this one and I'll bring in #99716 on top of it.

ChrisHegarty committed:
remove superfluous super.close

ChrisHegarty committed:
Update comment regarding eventual removal of static methods

ChrisHegarty committed:
remove stray TODO comment

ChrisHegarty committed:
Add comment to BlockFactoryParameters

ChrisHegarty committed:
remove leftover

ChrisHegarty committed:
Update breaker in tests

ChrisHegarty committed:
Merge branch 'main' into blockFactory
